bin/comment_on_pr

```diff
@@ -1,14 +1,143 @@
 #!/bin/bash -eu
 
-PR_MESSAGE_FILE=$1
-GITHUB_TOKEN=$2
+# This script is used to comment on a pull request, and must be run in a buldkite PR step.
+#
+# Usage:
+#    [GITHUB_TOKEN=<token>] comment_on_pr [--id <comment-id>] [--if-exist update|delete] [--github-token <token>] <comment>
+#
+# Examples:
+#
+#    # Post a new comment which can be uniquely identified by the given comment id,
+#    # or update the existing comment that's associated with the given comment id.
+#    # By default, the existing comment will be updated. You can choose to delete
+#    # the existing comment by passing `--if-exist delete`.
+#    comment_on_pr --id comment-id [--if-exist update|delete] "This is a comment"
+#
+#    # Delete the comment associated with the given comment id by passing an empty comment
+#    comment_on_pr --id comment-id [--if-exist delete]
+#
+#    # Deprecated: Post a new comment to the PR
+#    comment_on_pr "This is a comment" <github-token>
+#
+# Please note the comment argument can be a path to a file containing the comment, or a markdown.
 
-if ! command -v gh &> /dev/null; then
-	brew install gh
+# Check dependencies and print their versions for diagnosis purposes
+cat <<EOF
+> jq --version
+$(jq --version)
+
+> curl --version
+$(curl --version)
+
+EOF
+
+if [[ ! "${BUILDKITE_PULL_REQUEST:-invalid}" =~ ^[0-9]+$ ]]; then
+    echo "Error: this tool can only be called from a Buildkite PR job"
+    exit 1
+fi
+
+# Default options
+opt_comment_id=""
+opt_if_exist="update"
+opt_github_token="${GITHUB_TOKEN:-}"
+
+# If there are two arguments and they don't start with an option, then we assume
+# this utility is called in the deprecated way: comment_on_pr <comment> <github-token>
+if [[ $# == 2 && "$1" != "--"* ]]; then
+    if [[ -f "$1" ]]; then
+        arg_pr_comment=$(cat "$1")
+    else
+        arg_pr_comment="$1"
+    fi
+
+    opt_github_token=$2
+else
+    while [[ "$#" -gt 0 ]]; do
+        case $1 in
+            --id) opt_comment_id="$2"; shift ;;
+            --if-exist) opt_if_exist="$2"; shift ;;
+            --github-token) opt_github_token="$2"; shift ;;
+            --*) echo "Unknown option: $1"; exit 1 ;;
+            *) break ;;
+        esac
+        shift
+    done
+
+    case "$#" in
+        0)
+            arg_pr_comment=""
+            ;;
+        1)
+            if [[ -f "$1" ]]; then
+                arg_pr_comment=$(cat "$1")
+            else
+                arg_pr_comment="$1"
+            fi
+            ;;
+        *) echo "Error: too many arguments"; exit 1 ;;
+    esac
+fi
+
+export GITHUB_TOKEN="$opt_github_token"
+
+github_user=$(basename "$(dirname "${BUILDKITE_PULL_REQUEST_REPO%.git}")")
+github_repo=$(basename "${BUILDKITE_PULL_REQUEST_REPO%.git}")
+issues_endpoint="repos/${github_user}/${github_repo}/issues"
+
+all_pr_comments=$(
+    github_api "$issues_endpoint/${BUILDKITE_PULL_REQUEST}/comments" --fail \
+    || (>&2 echo "Error: The GitHub token is invalid" && exit 1)
```

Comment:
🛑  The `|| (…)` creates a subshell, so your `exit 1` inside this only exits the subshell within `(…)` instead of exiting the main script.

I had a doubt if this was the case or not so I tested it in a small shell and confirmed:
```bash
#!/bin/bash

echo "Hello"
false || (>&2 echo "Error" && exit 1)
echo "Bye"
```
```
$ ./test.sh 
Hello
Error
Bye
```

Reply:
Also, are we sure that the only possible reason for the request to fail would always be because "The GitHub token is invalid"?

Maybe it would make more sense to move the logic of `curl` response check into the `github_api` utility instead, and make that utility check for the `$?` exit status of `curl` and if it's non-zero, print the error output of the request in `stderr`… Though… now that I think about it I'm expecting `curl … --fail` to do that (print request error message on stderr) already, so maybe all there's to do is `github_api … --fail-with-body || exit 1` here or something like that? 🤔

Reply:
This utility has the `-e` flag, which would make the `exit` statement work. I see the expected result on my end if add `-e` option to your `test.sh` script.

> so maybe all there's to do is github_api … --fail-with-body || exit 1 here or something like that?

I don't think we can do that if we'd like to capture the API response to be used later. Because curl prints "transportation errors" on stderr, but prints the HTTP response errors (4xx and 5xx for example) to stdout, which means it would be captured by the `all_pr_comments` variable.

Reply:
> are we sure that the only possible reason for the request to fail would always be because "The GitHub token is invalid"?

Definitely not the only reason. Given this script is suppose to be called in PR jobs. The curl errors could be:
1. GitHub API is down
2. The GitHub token is not a token at all.
3. The GitHub token is correct, but do not have permission to look PR comments in the repo.

I have updated the error message with more details in https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/51/commits/b9104aa7bcfca34369c54fbbf100aceff1f459cf.

Reply:
Then maybe the `github_api` utility needs to handle the "capture curl's stdout, then if `$? -eq 0` then just return the fairies string… but if non-zero then print the captured error on `stderr` instead, and forward the exit code"?

Reply:
It just occurred to me that I can use `--output /dev/stderr` to print the HTTP errors to stderr. See https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/51/commits/7c45813bf76b5cfca16742d33cb24221da08dd65.

Reply:
I don't want the `github_api` utility to do too much. `curl` command has loads of options to tweak its behaviour. It's probably more flexible to let the user decides what they want.

Reply:
Does that still capture the result in `all_pr_comments` in case of success? Wouldn't that print the request's valid JSON response to `stderr` too even if no error, and thus not have it captured into `all_pr_comments` as expected?

Reply:
> I don't want the github_api utility to do too much. curl command has loads of options to tweak its behaviour. It's probably more flexible to let the user decides what they want.

If there's already a set of curl options we can use to tell it "print output to stdout on success but to stderr on failure" (so that valid requests' responses get captured in `$all_pr_comments` on success but error is printed to stderr, instead of captured, on error) then I'm all for it!

Reply:
Ah! Check [this](https://stackoverflow.com/a/18284752)! 🙂

Reply:
I re-worked this part a little bit in https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/51/commits/add969cf7cb55dfa7f016607d7ef0c233dd91106 and also addressed [your concern around escaping CLI arguments](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/51#discussion_r1190923791). Let me know what you think.

Reply:
That looks perfect, thanks! 👌